pkg/relayer/proxy/server_builder.go

```diff
@@ -42,17 +41,16 @@ func (rp *relayerProxy) BuildProvidedServices(ctx context.Context) error {
 		var serviceEndpoints []relayer.RelayServer
 
 		for _, endpoint := range serviceConfig.Endpoints {
-			url, err := url.Parse(endpoint.Url)
-			if err != nil {
-				return err
-			}
-			supplierEndpointHost := url.Host
+			// TODO: Move `supplierEndpointHost` into a separate config file (could be a part of chains.json if we
```

Comment:
Should this TODO mention #137? I'm confused because #137 seems to be addressing stake, but I am not sure about the configuration of the service itself.

Reply:
> Should this TODO mention #137?

Yes. 

> I'm confused because #137 seems to be addressing stake, but I am not sure about the configuration of the service itself.

When you run stake you say: `I, supplier #1, am staking for 100 POKT for service X with endpoint XX and service Y with endpoint YY, etc...

Here's an example:

![Screenshot 2023-11-22 at 12 15 03 PM](https://github.com/pokt-network/poktroll/assets/1892194/44615a9d-b0e6-411e-b363-6a98908cb77b)

Reply:
After reviewing the situation, I've realized there are several issues with our current configuration setup that I initially overlooked, leading to some confusion. Here's a summary of my observations and suggestions:

The configuration file we're using is named 'stake' config, which seems misleading. In the mainnet environment, 'stake' typically refers to components like the node's discoverable endpoint (the HTTPS endpoint for /v1/relay) and the supported services (chains). However, this 'stake' config we have is more of an internal setup specific to supplier1 and doesn't necessarily reflect the services staked on the network (e.g. current chains.json). Therefore, I believe we should consider renaming it to avoid confusion. Perhaps, there could be a separate file that can specify the services supported by the node, and the URL the node can be reached at?

I encountered a problem where the endpoint defined in this config file was being used to bind or open a port. For instance, if the config file specifies an 'anvil' hostname, the supplier attempts to open a port on this hostname, which is impractical unless there's a network interface corresponding to the 'anvil' DNS. My current workaround, setting a hardcoded '0.0.0.0:8545', is a temporary fix and won't support multiple endpoints. We should consider relocating this endpoint to either another config file or a command-line argument to make it more dynamic. I can explore this further in a separate pull request to address the immediate issue.

I admit that the overall configuration, port management, and service interactions are a bit confusing to me. To tackle this, I've started compiling information on a [Notion page](https://www.notion.so/buildwithgrove/Actors-configs-dependencies-etc-ab9a93322b8744a2aaa089d5b4758b75?pvs=4) which outlines our system's actors, configurations, and dependencies. This effort should not only clarify things for me but also serve as a valuable resource for #159 .

Reply:
@okdas I believe this relates to the [comment](
https://github.com/pokt-network/poktroll/issues/205#issuecomment-1822094055) I left on @red-0ne's PR in #205:

Staking configs:
- `App` config -> for `staking` applications
- `Gateway` config -> for `staking` gateways
- `Supplier` config -> for `staking` suppliers

Operations configs:
- `AppGate` config -> for `operating` a gateway or an application
- `RelayMiner` config -> for `operating` a supplier

Reply:
> After reviewing the situation, I've realized there are several issues with our current configuration setup that I initially overlooked, leading to some confusion. Here's a summary of my observations and suggestions:

1. Is this a blocker for this PR or subsequent ones?

> I admit that the overall configuration, port management, and service interactions are a bit confusing to me. To tackle this, I've started compiling information on a [Notion page](https://www.notion.so/buildwithgrove/Actors-configs-dependencies-etc-ab9a93322b8744a2aaa089d5b4758b75?pvs=4) which outlines our system's actors, configurations, and dependencies. This effort should not only clarify things for me but also serve as a valuable resource for https://github.com/pokt-network/poktroll/issues/159 .

2. I agree and have discussed it with @red-0ne. Documenting them is a top priority.

> The configuration file we're using is named 'stake' config, which seems misleading. In the mainnet environment, 'stake' typically refers to components like the node's discoverable endpoint (the HTTPS endpoint for /v1/relay) and the supported services (chains). However, this 'stake' config we have is more of an internal setup specific to supplier1 and doesn't necessarily reflect the services staked on the network (e.g. current chains.json). Therefore, I believe we should consider renaming it to avoid confusion. Perhaps, there could be a separate file that can specify the services supported by the node, and the URL the node can be reached at?

3. 

We `STAKE` the `App`, `Gateway` and `Supplier.
We `DEPLOY` the `AppGate` and `RelayMiner`

As of writing this comment, the state on main is (I manually added space and comments to the output):

```bash
$ find . -name '*.yaml' | grep "config" | sort

./localnet/poktrolld/config/appgate_server_config.yaml # Deploy Appgate (internal, off-chain)

./localnet/poktrolld/config/application1_stake_config.yaml # Stake App (external, on-chain)
./localnet/poktrolld/config/application2_stake_config.yaml # Stake App (external, on-chain)
./localnet/poktrolld/config/application3_stake_config.yaml # Stake App (external, on-chain)

./localnet/poktrolld/config/relayminer_config.yaml  deploy RelayMiner (internal, off-chain)

./localnet/poktrolld/config/supplier1_stake_config.yaml # stake supplier (external, on-chain)
./localnet/poktrolld/config/supplier2_stake_config.yaml # stake supplier (external, on-chain)
./localnet/poktrolld/config/supplier3_stake_config.yaml # stake supplier (external, on-chain)

./localnet_config.yaml
```

As you can see the only thing we're missing is `Stake Gateway`

> I encountered a problem where the endpoint defined in this config file was being used to bind or open a port. For instance, if the config file specifies an 'anvil' hostname, the supplier attempts to open a port on this hostname, which is impractical unless there's a network interface corresponding to the 'anvil' DNS. My current workaround, setting a hardcoded '0.0.0.0:8545', is a temporary fix and won't support multiple endpoints. We should consider relocating this endpoint to either another config file or a command-line argument to make it more dynamic. I can explore this further in a separate pull request to address the immediate issue.

4. 

If you look at `localnet/poktrolld/config/supplier1_stake_config.yaml` for example. The expectation here is that when you (as an app) get paired with `supplier1` during a session, you'd be using `http://anvil:8547` to request service  for `anvil`. In practice, this would be a domain like `anvil.supplier1.com` behind a load balancer. 

```yaml
- service_id: anvil
  endpoints:
  - url: http://anvil:8547
    rpc_type: json_rpc
- service_id: svc1
  endpoints:
  - url: http://lovalhost:8081
    rpc_type: json_rpc 
```

@okdas I'd defer to you for the best way to handle this right now, but I'm assuming we should (in theory) something like the following:

Replace `http://anvil:8547` with `anvil.supplier1.poktroll.com` and update DNS configs for `poktroll.com` to be different in localnet/devnet.

_Side note: Looks like we have a typo of `lovalhost` to `localhost`._

Reply:
@okdas , `supplier1_stake_config.yaml` should be used to stake and advertise endpoints on-chain.

The fact that we have `anvil` is definitely misleading. Since what the supplier should be advertising is the service and not a specific implementation of it.

I propose having `supplier1_stake_config.yaml` with the following:

```yaml
- service_id: ethereum
  endpoints:
  - url: http://ethereum.supplier1.poktroll.com:<whatever_port>
    rpc_type: json_rpc
```

This file will be used to stake and have the previous config available on-chain. So, for bootstrapping the chain, I would recommend issuing a `poktrolld tx supplier-stake 1000upokt --config supplier1_stake_config.yaml` instead of having hardcoded values in genesis configuration `config.yaml`.

So, when starting the relay miner, the `RelayerProxy component` would query its on-chain supplier entry and make the relevant services available.

Just like it's done here:
```go
for _, endpoint := range serviceConfig.Endpoints {
	url, err := url.Parse(endpoint.Url)
	if err != nil {
		return err
	}
	supplierEndpointHost := url.Host

	var server relayer.RelayServer

	switch endpoint.RpcType {
	case sharedtypes.RPCType_JSON_RPC:
		server = NewSynchronousServer(
			service,
			supplierEndpointHost,
			proxiedServicesEndpoints,
			rp.servedRelaysPublishCh,
			rp,
		)
	default:
		return ErrRelayerProxyUnsupportedRPCType
	}
	serviceEndpoints = append(serviceEndpoints, server)
}
```

Which should be reverted in the current commit BTW https://github.com/pokt-network/poktroll/pull/198/files#diff-360f9ba4c864cfdbf1bd894500d507361d5c04c9d78d064d2062140dcdbaf62bR47